Error cleaning up after pivot: remove /var/lib/docker/tmp/docker-builder113372687/.pivot_root936443744: no such file or directory

**Output of `docker version`:**

```
 Version:      1.12.0-dev
 API version:  1.24
 Go version:   go1.6.1
 Git commit:   08ec360
 Built:        Sun May  8 12:39:42 2016
 OS/Arch:      linux/amd64
 Experimental: true

Server:
 Version:      1.12.0-dev
 API version:  1.24
 Go version:   go1.6.1
 Git commit:   08ec360
 Built:        Sun May  8 12:39:42 2016
 OS/Arch:      linux/amd64
 Experimental: true
```


**Output of `docker info`:**

```
docker info
Containers: 30
 Running: 0
 Paused: 0
 Stopped: 30
Images: 138
Server Version: 1.12.0-dev
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: journald
Cgroup Driver: systemd
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.5.2-302.fc24.x86_64
Operating System: Fedora 24 (Workstation Edition)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 11.44 GiB
Name: localhost.localdomain
ID: JQDM:ZDU3:BYSR:SIMJ:QGDH:GBGW:MF7Q:VYLQ:UYFD:W5KY:H4WG:PIMF
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 13
 Goroutines: 25
 System Time: 2016-05-08T12:42:03.783123312+02:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 localhost
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
physical


**Steps to reproduce the issue:**
1. `make`
2.
3.


**Describe the results you received:**
```
$ make
docker build  -t "docker-dev:master" -f "Dockerfile" .
Sending build context to Docker daemon 196.1 MB
Error response from daemon: Untar re-exec error: exit status 1: output: Error cleaning up after pivot: remove /var/lib/docker/tmp/docker-builder606958863/.pivot_root892405879: no such file or directory
Makefile:67: recipe for target 'build' failed
make: *** [build] Error 1**
```

**Describe the results you expected:**
`make` works

**Additional information you deem important (e.g. issue happens only occasionally):**
Reverting https://github.com/docker/docker/pull/22506 makes this working again.

--- a/pkg/chrootarchive/chroot_linux.go
+++ b/pkg/chrootarchive/chroot_linux.go
@@ -43,7 +43,9 @@ func chroot(path string) (err error) {
 		}
 
 		errCleanup := os.Remove(pivotDir)
-		if errCleanup != nil {
+		// pivotDir doesn't exist if pivot_root failed and chroot+chdir was successful
+		// but we already cleaned it up on failed pivot_root
+		if errCleanup != nil && !os.IsNotExist(errCleanup) {
 			errCleanup = fmt.Errorf("Error cleaning up after pivot: %v", errCleanup)
 			if err == nil {
 				err = errCleanup
@@ -52,7 +54,10 @@ func chroot(path string) (err error) {
 	}()
 
 	if err := syscall.PivotRoot(path, pivotDir); err != nil {
-		// If pivot fails, fall back to the normal chroot
+		// If pivot fails, fall back to the normal chroot after cleaning up temp dir for pivot_root
+		if err := os.Remove(pivotDir); err != nil {
+			return fmt.Errorf("Error cleaning up after failed pivot: %v", err)
+		}
 		return realChroot(path)
 	}
 	mounted = true
@@ -84,7 +89,7 @@ func realChroot(path string) error {
 		return fmt.Errorf("Error after fallback to chroot: %v", err)
 	}
 	if err := syscall.Chdir("/"); err != nil {
-		return fmt.Errorf("Error chaning to new root after chroot: %v", err)
+		return fmt.Errorf("Error changing to new root after chroot: %v", err)
 	}
 	return nil
 }

